ouroboros-network/src/Ouroboros/Network/Subscription/Dns.hs

```diff
@@ -0,0 +1,265 @@
+    }
+
+dnsResolve :: forall m.
+     ( MonadAsync m
+     , MonadSay   m
+     , MonadSTM   m
+     , MonadTime  m
+     , MonadTimer m
+     , MonadThrow m
+     )
+    => Tracer m DnsTrace
+    -> Resolver m
+    -> DnsSubscriptionTarget
+    -> m (SubscriptionTarget m Socket.SockAddr)
+dnsResolve tracer resolver (DnsSubscriptionTarget domain _ _) = do
+    ipv6Rsps <- newEmptyTMVarM
+    ipv4Rsps <- newEmptyTMVarM
+    gotIpv6Rsp <- newTVarM False
+
+    aid_ipv6 <- async $ resolveAAAA gotIpv6Rsp ipv6Rsps
+    aid_ipv4 <- async $ resolveA gotIpv6Rsp ipv4Rsps
+
+    rd_e <- waitEitherCatch aid_ipv6 aid_ipv4
+    case rd_e of
+         Left r_ipv6 -> -- AAAA lookup finished first
+             case r_ipv6 of
+                  Left e_ipv6 -> do -- AAAA lookup failed
+                      traceWith tracer $ DnsTraceLookupException e_ipv6
+                      return $ SubscriptionTarget $ listTargets (Right ipv4Rsps) (Left [])
+                  Right _ -> do
+                      -- Try to use IPv6 result first.
+                      traceWith tracer DnsTraceLookupIPv6First
+                      ipv6Res <- atomically $ takeTMVar ipv6Rsps
+                      return $ SubscriptionTarget $ listTargets (Left ipv6Res) (Right ipv4Rsps)
+         Right r_ipv4 ->
+              case r_ipv4 of
+                  Left e_ipv4 -> do -- A lookup failed
+                      traceWith tracer $ DnsTraceLookupException e_ipv4
+                      return $ SubscriptionTarget $ listTargets (Right ipv6Rsps) (Left [])
+                  Right _ -> do
+                      traceWith tracer DnsTraceLookupIPv4First
+                      return $ SubscriptionTarget $ listTargets (Right ipv4Rsps) (Right ipv6Rsps)
+
+  where
+
+    {-
+     - Returns a series of SockAddr, where the address family will alternate
+     - between IPv4 and IPv6 as soon as the corresponding lookup call has completed.
+     -}
+    listTargets :: Either [Socket.SockAddr] (TMVar m [Socket.SockAddr])
+                -> Either [Socket.SockAddr] (TMVar m [Socket.SockAddr])
+                -> m (Maybe (Socket.SockAddr, SubscriptionTarget m Socket.SockAddr))
+
+    -- No result left to try
+    listTargets (Left []) (Left []) = return Nothing
+
+    -- No results left to try for one family
+    listTargets (Left []) ipvB = listTargets ipvB (Left [])
+
+    -- Result for one address family
+    listTargets (Left (addr : addrs)) ipvB =
+        return $ Just (addr, SubscriptionTarget (listTargets ipvB (Left addrs)))
+
+    -- No result for either family yet.
+    listTargets (Right addrsVarA) (Right addrsVarB) = do
+        addrsRes <- atomically $ do
+            a_m <- tryReadTMVar addrsVarA
+            b_m <- tryReadTMVar addrsVarB
+            case (a_m, b_m) of
+                 (Nothing, Nothing) -> retry
+                 (Just a, _)        -> return $ Left a
+                 (_, Just b)        -> return $ Right b
```

Comment:
I think this could be simplified using `orElse`, e.g.
```
          (Left  <$> readTMVar addrsVarA)
 `orElse` (Right <$> readTMVar addrsVarB)
```
This will only block if both block.

Reply:
That can be done after the `orElse` MonadSTM pr is merged.

Reply:
Oh yes, fair enough since you're paramatric in `MonadSTM` here. So just leave a TODO with the `orElse` version so we come back to it.

Reply:
Will add TODO.